pkg/cloud/openstack/services/certificates/certificates.go

```diff
@@ -0,0 +1,334 @@
+/*
+Copyright 2018 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package certificates
+
+import (
+	"crypto/rand"
+	"crypto/rsa"
+	"crypto/sha256"
+	"crypto/x509"
+	"crypto/x509/pkix"
+	"encoding/hex"
+	"encoding/pem"
+	"fmt"
+	"github.com/pkg/errors"
+	"k8s.io/client-go/tools/clientcmd/api"
+	"k8s.io/klog"
+	"math"
+	"math/big"
+	"net"
+	"sigs.k8s.io/cluster-api-provider-openstack/pkg/apis/openstackproviderconfig/v1alpha1"
+	"strings"
+	"time"
+)
+
+const (
+	rsaKeySize     = 2048
+	duration365d   = time.Hour * 24 * 365
+	clusterCA      = "cluster-ca"
+	etcdCA         = "etcd-ca"
+	frontProxyCA   = "front-proxy-ca"
+	serviceAccount = "service-account"
+)
+
+// NewPrivateKey creates an RSA private key
+func NewPrivateKey() (*rsa.PrivateKey, error) {
+	return rsa.GenerateKey(rand.Reader, rsaKeySize)
+}
+
+// AltNames contains the domain names and IP addresses that will be added
+// to the API Server's x509 certificate SubAltNames field. The values will
+// be passed directly to the x509.Certificate object.
+type AltNames struct {
+	DNSNames []string
+	IPs      []net.IP
+}
+
+// Config contains the basic fields required for creating a certificate
+type Config struct {
+	CommonName   string
+	Organization []string
+	AltNames     AltNames
+	Usages       []x509.ExtKeyUsage
+}
+
+// ReconcileCertificates generate certificates if none exists.
+func (s *Service) ReconcileCertificates(clusterName string, clusterProviderSpec *v1alpha1.OpenstackClusterProviderSpec) error {
+	if !clusterProviderSpec.CAKeyPair.HasCertAndKey() {
```

Comment:
@jichenjc Your comment from the PR on the other repository:
> from code, looks like you are trying to generate those keys? if it's generated, then how do you store them and reuse? looks to me the Spec are the desired state but if you didn't give key at beginning, it's not desired state?? just curious ..

Reply:
@jichenjc The logic is the following:

If no certificates are set in the Cluster CRD, ReconcileCertificates will generate them and add them to the Cluster CRD. But it's also possible (if wanted) to set the certificates when creating the Cluster CRD.

The main reasoning behind this is that we need a place to store certificates so we can give the same certifcates to multiple control plane nodes, without copying them from one host to another. Also it's now possible to generate a kubeconfig based on the Cluster CRD alone.

Reply:
I understand the logic here :), the question I have originally is I didn't see some code like 'save' ' store' the Cert and key ...so I assume kubeconfig will be the place to hold those stuff..

Reply:
Got it :). That's not really intuitive but the whole spec & status are saved here: https://github.com/kubernetes-sigs/cluster-api-provider-openstack/blob/51c65e2000f3fa0a12931c8213ae1131a5878798/pkg/cloud/openstack/cluster/actuator.go#L107

But I see the `defer` statement is way to late. I moved it up in another PR. But I'll do it now in this PR.

Reply:
It should only be a matter of cluster actuator reconciling the certs. As soon as that's done the machine actuator can use them. As long as the certs are not in the Cluster CRD yet the machine actuator has to wait (if not it will render empty certs in the userData)